flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala

```diff
@@ -172,6 +172,19 @@ class ExpressionReducer(
             }
             reducedValues.add(maySkipNullLiteralReduce(rexBuilder, value, unreduced))
             reducedIdx += 1
+          case SqlTypeName.TIMESTAMP =>
+            val reducedValue = reduced.getField(reducedIdx)
+            val value = if (reducedValue != null) {
+              val ts = reducedValue.asInstanceOf[SqlTimestamp]
+              val milliseconds = ts.getMillisecond
+              val nanoseconds = ts.toLocalDateTime.getNano
```

Comment:
Why not just use `LocalDateTime` to `TimestampString`?

Reply:
A little tricky here, take `CAST('1500-04-30 00:00:00' AS TIMESTAMP(3))` as an example:
we use `SqlDateTimeUtils.toTimestamp`[1] to cast the string to SqlTimestamp and use `ExpressionReducer`[2] to make a Timestamp literal. 

The [1] step considers Gregorian cutovers and again the [2] step uses `TimestampString.fromMillisSinceEpoch` which calls our copied `DateTimeUtils.julianToString` considers Gregorian cutovers too. Two wrongs make a correct result.

I will change the `toTimestamp` and this `ExpressionReducer` to ignore Gregorian cutovers in next PR and that will be the final one.

Reply:
I think you can fix it, and it is very worthwhile to fix. Otherwise it will bring a lot of strange code.